pandas/core/generic.py

```diff
@@ -3796,6 +3902,13 @@ def __delitem__(self, key) -> None:
     # ----------------------------------------------------------------------
     # Unsorted
 
+    def _check_inplace_and_allows_duplicate_labels(self, inplace):
+        if inplace and not self.flags.allows_duplicate_labels:
+            raise ValueError(
```

Comment:
does this have reasonable testing?

Reply:
I have tests for every method it's called it (NDFrame.set_axis, NDFrame,rename, DataFrame.set_index, DataFrame.reset_index).

The potential issue is places where we do an inplace op that gets us in an invalid state (an NDFrame with duplicate labels but `allows_duplicate_labels=False`). So the known cases are tested, but there's possibly additional places where we need to check this. I hopefully got them all though.